c10/core/CPUCachingAllocator.cpp

```diff
@@ -0,0 +1,120 @@
+#include <c10/core/CPUCachingAllocator.h>
+
+namespace c10 {
+
+namespace {
+// We can make this thread_local perhaps. Then we would not need mutex.
+// However making it thread_local may increase memory usage.
+static CPUCachingAllocator cpu_caching_allocator;
+
+thread_local CachingAllocatorInfo caching_allocator_info;
+} // namespace
+
+CPUCachingAllocator& GetCPUCachingAllocator() {
+  return cpu_caching_allocator;
+}
+
+inline void* CPUCachingAllocator::allocate_and_cache(const size_t bytes) {
+  void* ptr;
```

Comment:
Possible future optimization: if there's some way query the underlying allocator for the buffer size that will actually be returned when requesting an allocation of size `bytes`, we can use that here and in `use_cached`.  That way, a cached allocation for 1023 bytes can (most likely) be used to satisfy a request for 1024 bytes.

Reply:
So CUDACaching allocator does this by using binary tree for allocation, but that adds some latency to allocations. We dont know if this is significant. We probably should add some profiling to confirm if this is the case or not. My intuition is that for static graph like networks allocation patterns are same. The optimization you mentioned can help us reduce memory footprint though.